File: scripts/update-defect-log.js
#!/usr/bin/env node
/**
 * update-defect-log.js
 *
 * Upserts a defect entry in docs/defect-log.json for a production bug.
 * Called by .github/workflows/auto-regression.yml
 *
 * Environment variables:
 *   ISSUE_NUMBER         - GitHub issue number
 *   ISSUE_TITLE          - GitHub issue title
 *   ISSUE_URL            - GitHub issue HTML URL
 *   AREA                 - Test group area
 *   REQ_IDS              - Comma-separated requirement IDs
 *   TEST_FILE            - Path to the generated test file (optional)
 *   REGRESSION_PR        - PR number to record (optional, set after PR creation)
 */

'use strict';

const fs = require('fs');

const LOG_FILE   = 'docs/defect-log.json';
const issueNumber = parseInt(process.env.ISSUE_NUMBER || '0', 10);
const now         = new Date().toISOString();

// Read existing log
let log = [];
if (fs.existsSync(LOG_FILE)) {
  try {
    log = JSON.parse(fs.readFileSync(LOG_FILE, 'utf8'));
  } catch (err) {
    console.error(`Warning: Could not parse ${LOG_FILE}:`, err.message);
    log = [];
  }
}

const idx = log.findIndex(e => e.issue_number === issueNumber);

if (idx >= 0) {
  // Update existing entry
  if (process.env.TEST_FILE) {
    log[idx].regression_test_created     = true;
    log[idx].regression_test_file        = process.env.TEST_FILE;
    log[idx].regression_test_created_at  = now;
  }
  if (process.env.REGRESSION_PR) {
    log[idx].regression_pr = parseInt(process.env.REGRESSION_PR, 10);
  }
} else {
  // Insert new entry
  log.push({
    issue_number:                issueNumber,
    issue_title:                 process.env.ISSUE_TITLE   || '',
    issue_url:                   process.env.ISSUE_URL     || '',
    area:                        process.env.AREA          || '',
    req_ids:                     (process.env.REQ_IDS || '')
                                   .split(',').map(r => r.trim()).filter(Boolean),
    logged_at:                   now,
    regression_test_created:     Boolean(process.env.TEST_FILE),
    regression_test_file:        process.env.TEST_FILE || null,
    regression_test_created_at:  process.env.TEST_FILE ? now : null,
    regression_pr:               process.env.REGRESSION_PR
                                   ? parseInt(process.env.REGRESSION_PR, 10)
                                   : null,
  });
}

fs.mkdirSync('docs', { recursive: true });
fs.writeFileSync(LOG_FILE, JSON.stringify(log, null, 2) + '\n', 'utf8');
console.log(`✅ Updated ${LOG_FILE}`);

File: scripts/update-defect-log.py
#!/usr/bin/env python3
"""
update-defect-log.py — Update docs/defect-log.json with regression PR details.

Usage:
  python3 scripts/update-defect-log.py \
    --issue-number 123 \
    --pr-number 456 \
    --test-file tests/playwright-agents/regression/issue-123-slug.spec.ts
"""
import argparse
import json
import os
import sys


def main():
    parser = argparse.ArgumentParser(description="Update defect-log.json with regression PR")
    parser.add_argument("--issue-number", required=True, type=int)
    parser.add_argument("--pr-number",    default="")
    parser.add_argument("--test-file",    default="")
    parser.add_argument("--log-path",     default="docs/defect-log.json")
    args = parser.parse_args()

    log_path = args.log_path

    try:
        with open(log_path) as f:
            log = json.load(f)
    except Exception:
        log = []

    pr_number = int(args.pr_number) if args.pr_number.isdigit() else None
    updated   = False

    for entry in log:
        if entry.get("issue") == args.issue_number:
            entry["regression_test_created"] = True
            entry["regression_pr"]           = pr_number
            entry["regression_test_file"]    = args.test_file
            updated = True
            break

    if not updated:
        # Entry may not exist yet (race condition with defect-tracker)
        log.append({
            "issue":                    args.issue_number,
            "title":                    "",
            "detected":                 "",
            "attributed_commit":        "unknown",
            "affected_area":            "unknown",
            "affected_url":             "",
            "req_ids":                  [],
            "regression_test_created":  True,
            "regression_pr":            pr_number,
            "regression_test_file":     args.test_file,
        })

    with open(log_path, "w") as f:
        json.dump(log, f, indent=2)
        f.write("\n")

    print(f"Updated {log_path} for issue #{args.issue_number}")


if __name__ == "__main__":
    main()
